tests/github/experimental/workflow/FederatedRuntime/testcase_private_attributes/workspace/testflow_privateattributes.ipynb

```diff
@@ -0,0 +1,411 @@
+    "        validate_collab_private_attrs(\n",
+    "            self, \"collaborator_step_b\", aggr=[\"test_loader_agg\"], collabs=[\"train_loader\", \"test_loader\"]\n",
+    "        )\n",
+    "        self.exclude_collab_to_agg = 10\n",
+    "        self.include_collab_to_agg = 12\n",
+    "        self.next(self.join, exclude=[\"exclude_collab_to_agg\"])\n",
+    "\n",
+    "    @aggregator\n",
+    "    def join(self, inputs):\n",
+    "        \"\"\"\n",
+    "        Testing whether attributes are excluded from collab to agg\n",
+    "        \"\"\"\n",
+    "        # Aggregator should only be able to access its own attributes\n",
+    "        if hasattr(self, \"test_loader_agg\") is False:\n",
+    "            TestFlowPrivateAttributes.ERROR_LIST.append(\n",
+    "                \"aggregator_join_aggregator_attributes_missing\"\n",
+    "            )\n",
+    "            print(\n",
+    "                f\"{bcolors.FAIL} ... Attribute test failed in join - aggregator private attributes\"\n",
+    "                + f\" not accessible {bcolors.ENDC}\"\n",
+    "            )\n",
+    "\n",
+    "        for input in enumerate(inputs):\n",
+    "            collab = input[1].input\n",
+    "            if (\n",
+    "                hasattr(input, \"train_loader\")\n",
+    "                or hasattr(input, \"test_loader\")\n",
+    "            ):\n",
+    "                # Error - we are able to access collaborator attributes\n",
+    "                TestFlowPrivateAttributes.ERROR_LIST.append(\n",
+    "                    \"join_collaborator_attributes_found\"\n",
+    "                )\n",
+    "                print(\n",
+    "                    f\"{bcolors.FAIL} ... Attribute test failed in Join - Collaborator: {collab}\"\n",
+    "                    + f\" private attributes accessible {bcolors.ENDC}\"\n",
+    "                )\n",
+    "\n",
+    "        self.next(self.end)\n",
+    "\n",
+    "    @aggregator\n",
+    "    def end(self):\n",
+    "        \"\"\"\n",
+    "        This is the 'end' step. All flows must have an 'end' step, which is the\n",
+    "        last step in the flow.\n",
+    "\n",
+    "        \"\"\"\n",
+    "        print(\n",
+    "            f\"{bcolors.OKBLUE}Testing FederatedFlow - Ending Test  for accessibility of private \"\n",
+    "            + f\"attributes  {bcolors.ENDC}\"\n",
+    "        )\n",
+    "\n",
+    "        if TestFlowPrivateAttributes.ERROR_LIST:\n",
+    "            raise (\n",
+    "                AssertionError(\n",
+    "                    f\"{bcolors.FAIL}\\n ...Test case failed ... {bcolors.ENDC}\"\n",
+    "                )\n",
+    "            )\n",
+    "        else:\n",
+    "            print(f\"{bcolors.OKGREEN}\\n ...Test case passed ... {bcolors.ENDC}\")\n",
+    "\n",
+    "        TestFlowPrivateAttributes.ERROR_LIST = []\n",
+    "\n",
+    "\n",
+    "def validate_agg_private_attr(self, step_name, **private_attrs_kwargs):\n",
```

Comment:
I thought the underlying framework didn't even support transferring private attributes over the network (as those could include non-serializable objects). Are there cases where the private attributes _could be_ transmitted, and therefore this test is needed?

Reply:
Private attributes are not transferred over the network, as they are already available to the Director and Envoy admins and are initialized in the Aggregator and collaborators during instantiation. This test case is necessary to confirm that private attributes of the Aggregator are not visible to the collaborator and vice versa. 
This ensures encapsulation and prevents unintended access between components.